service/graph_test.go

```diff
@@ -0,0 +1,193 @@
+	context.Context
+	sync.Mutex
+	next  int
+	order map[component.ID]int
+}
+
+func (c *contextWithOrder) record(id component.ID) {
+	c.Lock()
+	c.order[id] = c.next
+	c.next++
+	c.Unlock()
+}
+
+func TestGraphStartStop(t *testing.T) {
+	testCases := []struct {
+		name  string
+		edges [][2]component.ID
+	}{
+		{
+			name: "single",
+			edges: [][2]component.ID{
+				{component.NewIDWithName("r", "1"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("r", "2"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("p", "2")},
+				{component.NewIDWithName("p", "2"), component.NewIDWithName("e", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("e", "2")},
+			},
+		},
+		{
+			name: "multi",
+			edges: [][2]component.ID{
+				// Pipeline 1
+				{component.NewIDWithName("r", "1"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("r", "2"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("p", "2")},
+				{component.NewIDWithName("p", "2"), component.NewIDWithName("e", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("e", "2")},
+
+				// Pipeline 2, shares r1 and e2
+				{component.NewIDWithName("r", "1"), component.NewIDWithName("p", "3")},
+				{component.NewIDWithName("p", "3"), component.NewIDWithName("e", "2")},
+			},
+		},
+		{
+			name: "connected",
+			edges: [][2]component.ID{
+				// Pipeline 1
+				{component.NewIDWithName("r", "1"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("r", "2"), component.NewIDWithName("p", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("p", "2")},
+				{component.NewIDWithName("p", "2"), component.NewIDWithName("e", "1")},
+				{component.NewIDWithName("p", "1"), component.NewIDWithName("c", "1")},
+
+				// Pipeline 2, shares r1 and c1
+				{component.NewIDWithName("r", "1"), component.NewIDWithName("p", "3")},
+				{component.NewIDWithName("p", "3"), component.NewIDWithName("c", "1")},
+
+				// Pipeline 3, emits to e2 and c2
+				{component.NewIDWithName("c", "1"), component.NewIDWithName("e", "2")},
+				{component.NewIDWithName("c", "1"), component.NewIDWithName("c", "2")},
+
+				// Pipeline 4, also emits to e2
+				{component.NewIDWithName("c", "2"), component.NewIDWithName("e", "2")},
+			},
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &contextWithOrder{
+				Context: context.Background(),
+				order:   map[component.ID]int{},
+			}
+
+			pg := &pipelinesGraph{componentGraph: simple.NewDirectedGraph()}
+			for _, edge := range tt.edges {
+				f, t := &testNode{id: edge[0]}, &testNode{id: edge[1]}
+				pg.componentGraph.SetEdge(simple.Edge{F: f, T: t})
+			}
+
+			require.NoError(t, pg.StartAll(ctx, componenttest.NewNopHost()))
+			for _, edge := range tt.edges {
+				assert.Greater(t, ctx.order[edge[0]], ctx.order[edge[1]])
```

Comment:
To make sure we don't by mistake get this right, can we have a different map for start order and stop order?

Reply:
Good call. Now resetting the map before calling ShutdownAll.